drivers/adc/adc_w91.c

```diff
@@ -0,0 +1,343 @@
+/*
+ * Copyright (c) 2024 Telink Semiconductor (Shanghai) Co., Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT telink_w91_adc
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(adc_telink);
+
+#include <zephyr/drivers/adc.h>
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+
+#include "adc_context.h"
+#include <zephyr/drivers/pinctrl.h>
+#include <ipc/ipc_based_driver.h>
+#include <stdlib.h>
+
+#define ADC_REF 3300
+
+#define ADC_WORK_QUEUE_STACK_SIZE	512
+
+static struct k_work_q adc_work_queue;
+static K_KERNEL_STACK_DEFINE(adc_work_queue_stack,
+			     ADC_WORK_QUEUE_STACK_SIZE);
+
+enum {
+	IPC_DISPATCHER_ADC_SETUP = IPC_DISPATCHER_ADC,
+	IPC_DISPATCHER_ADC_READ,
+	IPC_DISPATCHER_ADC_IRQ_EVENT,
+};
+
+struct adc_w91_data {
+	struct adc_context ctx;
+	int16_t *buffer;
+	int16_t *repeat_buffer;
+	uint8_t channel;
+	struct ipc_based_driver ipc;
+	uint8_t instance_id;
+	struct k_work irq_cb_work;
+	const struct device *dev;
+};
+
+struct adc_read_req {
+	uint16_t ch;
+	uint32_t len;
+};
+
+struct adc_read_resp {
+	int err;
+	uint16_t adc_value;
+};
+
+/* ADC configuration structure */
+struct adc_w91_cfg {
+	uint16_t vref_internal_mv;
+	const struct pinctrl_dev_config *pcfg;
+	uint8_t instance_id;
+};
+
+IPC_DISPATCHER_PACK_FUNC_WITHOUT_PARAM(adc_w91_ipc_setup, IPC_DISPATCHER_ADC_SETUP);
+IPC_DISPATCHER_UNPACK_FUNC_ONLY_WITH_ERROR_PARAM(adc_w91_ipc_setup);
+
+static int adc_w91_ipc_setup(const struct device *dev)
+{
+	int err;
+
+	struct ipc_based_driver *ipc_data = &((struct adc_w91_data *)dev->data)->ipc;
+	uint8_t inst = ((struct adc_w91_cfg *)dev->config)->instance_id;
+
+	IPC_DISPATCHER_HOST_SEND_DATA(ipc_data, inst,
+		adc_w91_ipc_setup, NULL, &err,
+		CONFIG_TELINK_W91_IPC_DISPATCHER_TIMEOUT_MS);
+
+	return err;
+}
+
+/* APIs implementation: ADC read */
+static size_t pack_adc_w91_ipc_read(uint8_t inst, void *unpack_data, uint8_t *pack_data)
+{
+	struct adc_read_req *p_adc_read = unpack_data;
+	size_t pack_data_len =
+		sizeof(uint32_t) + sizeof(p_adc_read->ch) + sizeof(p_adc_read->len);
+
+	if (pack_data != NULL) {
+		uint32_t id = IPC_DISPATCHER_MK_ID(IPC_DISPATCHER_ADC_READ, inst);
+
+		IPC_DISPATCHER_PACK_FIELD(pack_data, id);
+		IPC_DISPATCHER_PACK_FIELD(pack_data, p_adc_read->ch);
+		IPC_DISPATCHER_PACK_FIELD(pack_data, p_adc_read->len);
+	}
+	return pack_data_len;
+}
+
+IPC_DISPATCHER_UNPACK_FUNC_ONLY_WITH_ERROR_PARAM(adc_w91_ipc_read);
+
+static int adc_w91_ipc_read(struct adc_w91_data *data, uint16_t ch,
+				   uint32_t len)
+{
+	struct adc_read_req adc_req = {.ch = ch, .len = len};
+	int err;
+
+	struct ipc_based_driver *ipc_data = &data->ipc;
+
+	IPC_DISPATCHER_HOST_SEND_DATA(ipc_data, data->instance_id, adc_w91_ipc_read, &adc_req, &err,
+				      CONFIG_TELINK_W91_IPC_DISPATCHER_TIMEOUT_MS);
+	if (err != 0) {
+		LOG_ERR("ADC read failed, ret(%d)", err);
+	}
+
+	return err;
+}
+
+/* Validate ADC data buffer size */
+static int adc_w91_validate_buffer_size(const struct adc_sequence *sequence)
```

Comment:
yes, this buffer can be too small

Reply:
Where is the buffer allocation?

Reply:
We declare adc_sequence stucture, for example, in ADC sample, adc_sequence contains buffer pointer and buffer size, so we should create this buffer separately.
	struct adc_sequence sequence = {
		.buffer = &buf,
		/* buffer size in bytes, not number of samples */
		.buffer_size = sizeof(buf),
	};
We can see the same buffer_size check in each vendor ADC implementation, because it wasn't implemented into Zephyr ADC APIs